qiskit/algorithms/optimizers/steppable_optimizer.py

```diff
@@ -0,0 +1,286 @@
+    https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/009_ask_and_tell.html
+
+
+    """
+
+    def __init__(
+        self,
+        maxiter: int = 1000,
+    ):
+        """
+        Args:
+            maxiter: Number of steps in the optimization process before ending the loop.
+        """
+        super().__init__()
+        self._state: OptimizerState = None
+        self.maxiter = maxiter
+
+    def ask(self) -> AskObject:
+        """Ask the optimizer for a set of points to evaluate.
+
+        This method asks the optimizer which are the next points to evaluate.
+        These points can, e.g., correspond to function values and/or its derivative.
+        It may also correspond to variables that let the user infer which points to evaluate.
+        It is the first method inside of a :meth:`~.step` in the optimization process.
+
+        Returns:
+            Since the way to evaluate the function can vary much with different
+            optimization algorithms, the object will be a custom dataclass for each optimizer.
+
+        """
+        raise NotImplementedError
+
+    def tell(self, ask_object: AskObject, tell_object: TellObject) -> None:
+        """Updates the optimization state once the objective function has been evaluated.
+        A canonical optimization example using :meth:`~.ask`
+        and :meth:`~.tell` can be seen in :meth:`~.step`.
+
+        Args:
+            ask_object: Contains the information on how the evaluation was done.
+            tell_object: Contains all relevant information about the evaluation of the objective
+                function.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def evaluate(self, ask_object: AskObject) -> TellObject:
+        """
+        Evaluates the function according to the instructions contained in :attr:`~.ask_object`.
+        If the user decides to use :meth:`~.step` instead of :meth:`~.ask` and :meth:`~.tell`
+        this function will contain the logic on how to evaluate the function.
+
+        Args:
+            ask_object: Contains the information on how to do the evaluation.
+
+        Return:
+            Contains all relevant information about the evaluation of the objective function.
+
+        """
+        raise NotImplementedError
+
+    def _callback_wrapper(self) -> None:
+        """
+        Wraps the callback function to accomodate each optimizer.
+        """
+        pass
+
+    def step(self) -> None:
+        """
+        Performs one step in the optimization process.
+        This method composes :meth:`~.ask`, :meth:`~.evaluate`, and :meth:`~.tell` to make a `step`
+        in the optimization process.
+        """
+        ask_object = self.ask()
+        tell_object = self.evaluate(ask_object=ask_object)
+        self.tell(ask_object=ask_object, tell_object=tell_object)
+
+    # pylint: disable=invalid-name
+    @abstractmethod
+    def initialize(
```

Comment:
What is disliked about this name - would it be better to pick something else `restart`, `reset_state` or whatever since I see these terms in the description below.

Reply:
If the user wants to do a step by step optimization, this method needs to be called at the very beginning of the process.

``` python
optimizer = GradienDescent()
optimizer.reset_state(x0=....)
while optimizer.continue_condition():
    optmizer.step()
result = optmizer.create_result()
```
Isn't it weird to call reset in this context?

Reply:
Maybe call it start() or something then? So you do start() then loop over step(). I assumed lint was saying it was not a good  name for a good reason.

Reply:
Okay, I understand it now. I will change it now. Thank you!